main.py

```diff
@@ -448,13 +475,401 @@ def move_methods_workaround(dex_filename, dex_filename_last, in_dir, out_dir, de
+    print_(" *** Decompressing services.jar ...")
+    decompress("services.jar", "services/")
+    #
+    # Disassemble it
+    print_(" *** Disassembling services classes...")
+    smali_to_search = "com/android/server/pm/PackageManagerService.smali"
+    smali_folder, dex_filename, dex_filename_last = find_smali(smali_to_search, "services/", DEVICE_SDK)
+    #
+    # Check the existence of the file to patch
+    if smali_folder is None:
+        print_(os.linesep+"ERROR: The services.jar classes file to patch cannot be found, please report the problem.")
+        exit_now(82)
+    to_patch = smali_folder+smali_to_search
+    #
+    # Do the injection
+    print_(" *** Patching services...")
+    f = open(to_patch, "r")
+    old_contents = f.readlines()
+    f.close()
+    #
+    f = open(SCRIPT_DIR+"/patches/allowsdwrite.smali", "r")
+    allowsdwrite = f.readlines()
+    f.close()
+    #
+    # services_classes - Add allowsdwrite method
+    #       for testing - \smali-classes\com\android\server\pm
+    i = 0
+    contents = []
+    already_patched = False
+    in_function = False
+    required_line_before = False
+    right_line = False
+    # start_of_line = None
+    done_patching = False
+    partially_patched = False
+    #
+    while i < len(old_contents):
+        if ";->allowSDWrite" in old_contents[i]:
+            already_patched = True
+        if ".method private allowSDWrite" in old_contents[i]:
+            partially_patched = True
+        if ".method private grantPermissionsLPw(Landroid/content/pm/PackageParser$Package;ZLjava/lang/String;)V" in old_contents[i]:
+            print_(" *** Detected: Android 7.x (or LOS/CM 14)")
+            in_function = True
+        if ".end method" in old_contents[i]:
+            in_function = False
+            required_line_before = False
+        if in_function and "const-wide/32 v26, 0x40000" in old_contents[i]:
+            required_line_before = True
+        if in_function and required_line_before and "invoke-static/range {v26 .. v27}" in old_contents[i]:
+            right_line = True
+            # start_of_line = i+1
+        if not already_patched and in_function and right_line and not done_patching:
+            # contents = contents[:start_of_line]
+            contents.append(old_contents[i])
+            contents.append("move-object/from16 v2, p0\n")
+            contents.append("move-object/from16 v3, p1\n")
+            contents.append("invoke-direct {v2, v3}, Lcom/android/server/pm/PackageManagerService;->allowSDWrite(Landroid/content/pm/PackageParser$Package;)V\n")
+            done_patching = True
+        else:
+            contents.append(old_contents[i])
+        i = i + 1
+    #
+    # Did we succeed with the Android 7.x patch?
+    if ((not done_patching) and (not partially_patched) and (not already_patched)):
+        #
+	# UNSUCCESSFUL - Retry alternative Android 6.x patch
+        print_(" *** Trying to detect Android 6.x (or LOS/CM 13)")
+	#
```

Comment:
__[flake8]__
* [W191] indentation contains tabs
* [E101] indentation contains mixed spaces and tabs

<!-- Comment by [SideCI](https://sideci.com) -->